### PR TITLE
feat: Add focusTargetOnClose prop to dialog component

### PR DESCRIPTION
**Pull request by scomea**
# Description
This change enables authors to specify an element to focus or callback to invoke when a dialog is unmounted.

## Motivation & context
https://github.com/microsoft/fast-dna/issues/2565

## Issue type checklist
- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [x] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

## Process & policy checklist
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast-dna/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

### CODE REVIEW

File: packages/fast-components-react-base/src/dialog/dialog.props.ts [hunk 1]
```diff
@@ -45,6 +45,14 @@ export interface DialogHandledProps extends DialogManagedClasses {
      */
     modal?: boolean;
 
+    /**
+     * Defines where focus goes when the dialog is unmounted.
+     * This is useful because modal dialogs will take focus back until that point in their life cycle.
+     * The element or ref provided will be focused when the dialog unmounts.
+     * Alternatively, providing a callback function gives developers more flexibility if needed.
+     */
+    refocusTarget?: React.RefObject<any> | HTMLElement | (() => void);
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/dialog/dialog.props.ts]: I think we should consider alternative names here so we don’t paint ourselves into a corner semantically. Refocus implies returning focus to a previously focused thing - there are valid scenarios where you may need to move focus to something which previously did not have focus. 

What about something along the lines of ‘focusTargetOnClose’, ‘focusTargetOnUnkount’, etc?

File: packages/fast-components-react-base/src/utilities/extractElement.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+        const foundNode: Element | Text | null = ReactDOM.findDOMNode(sourceRef.current);
+
+        if (foundNode instanceof HTMLElement) {
+            return foundNode;
+        }
+    }
+
+    return null;
+}
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/utilities/extractElement.ts]: Could this be useful elsewhere? It seems like since this is somewhat of a generic react utility that `fast-utilities-react` might be a better place for this. Would be great as well to get test coverage.

File: packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.spec.tsx [hunk 3]
```diff
@@ -1645,49 +1645,4 @@ describe("viewport positioner", (): void => {
...
-                    managedClasses={managedClasses}
-                />
-            </div>
-        );
-
-        const positioner: any = rendered.find("BaseViewportPositioner");
-        expect(positioner.instance()["extractElementFromRef"](anchorElement)).toBe(
-            anchorElement.current
-        );
-    });
```
Comment (chrisdholt) [inline packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.spec.tsx]: Should we do a separate PR to extract this utility? It seems significant enough to merit it's own PR.

File: packages/fast-components-react-base/src/dialog/dialog.props.ts (hunk 1 above)
Comment (nicholasrice) [inline packages/fast-components-react-base/src/dialog/dialog.props.ts]: What's the reason the Ref.RefObject can be `any`? Don't we want that to be an HTMLElement?

File: packages/fast-components-react-base/src/utilities/extractElement.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import ReactDOM from "react-dom";
+import { isNil } from "lodash-es";
+
+/**
+ * Attempts to extract an html element from a ref
+ */
+export function extractElementFromRef(
+    sourceRef: React.RefObject<any> | HTMLElement
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/utilities/extractElement.ts]: I think this should be `React.RefObject<HTMLElement>`

File: packages/fast-components-react-base/src/utilities/extractElement.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import ReactDOM from "react-dom";
+import { isNil } from "lodash-es";
+
+/**
+ * Attempts to extract an html element from a ref
+ */
+export function extractElementFromRef(
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/utilities/extractElement.ts]: I'd prefer to see a name more inline with the function implementation. This name implies it accepts a ref and returns an element, but it actually accepts a ref *or* an element. Can you either update the name to be more comprehensive or limit the scope of the function?

File: packages/fast-components-react-base/src/utilities/extractElement.ts (hunk 2 above)
Comment (scomea) [reply packages/fast-components-react-base/src/utilities/extractElement.ts]: See https://github.com/microsoft/fast-dna/pull/2613
Moved to the utilities in Base components and added tests,

File: packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.spec.tsx (hunk 3 above)
Comment (scomea) [reply packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.spec.tsx]: https://github.com/microsoft/fast-dna/pull/2613

File: packages/fast-components-react-base/src/dialog/dialog.props.ts (hunk 1 above)
Comment (scomea) [reply packages/fast-components-react-base/src/dialog/dialog.props.ts]: Updated, but moved to it's own pr
https://github.com/microsoft/fast-dna/pull/2613

File: packages/fast-components-react-base/src/utilities/extractElement.ts (hunk 4 above)
Comment (scomea) [reply packages/fast-components-react-base/src/utilities/extractElement.ts]: yep.  Moved to its own pr
https://github.com/microsoft/fast-dna/pull/2613

File: packages/fast-components-react-base/src/utilities/extractElement.ts (hunk 5 above)
Comment (scomea) [reply packages/fast-components-react-base/src/utilities/extractElement.ts]: Any suggestions?  I don't want to limit the scope as part of usefulness is in differentiating when a ref or element is passed in directly as we seem to be doing that in props with some regularity.

Comment (scomea) [reply packages/fast-components-react-base/src/utilities/extractElement.ts]: renamed to "extractHtmlElement" in the other PR

Comment (scomea) [reply packages/fast-components-react-base/src/utilities/extractElement.ts]: (if further discussion required let's do it there - https://github.com/microsoft/fast-dna/pull/2613)

File: packages/fast-components-react-base/src/dialog/dialog.props.ts (hunk 1 above)
Comment (scomea) [reply packages/fast-components-react-base/src/dialog/dialog.props.ts]: renamed to focusTargetOnClose

File: packages/fast-components-react-base/src/dialog/dialog.spec.tsx [hunk 6]
```diff
@@ -263,4 +263,37 @@ describe("dialog", (): void => {
 
         document.body.removeChild(container);
     });
+
+    test("should invoke focusTargetOnClose function on unmount of non-modal dialogs", () => {
```
Comment (nicholasrice) [inline packages/fast-components-react-base/src/dialog/dialog.spec.tsx]: Should we also test focus invocation of element / ref<htmlelement> props?

Comment (scomea) [reply packages/fast-components-react-base/src/dialog/dialog.spec.tsx]: added one for elements
